### PR TITLE
uniformization AS iAL output

### PR DESCRIPTION
**Pull request by calinbanc**
I've modified the output tables for the commands AS, iAL in order to uniform them.

### CODE REVIEW

File: security-analysis/src/main/java/com/powsybl/security/Security.java [hunk 1]
```diff
@@ -351,16 +575,186 @@ private static LimitViolationKey toKey(LimitViolation violation) {
...
+        };
+    }
+
+    private static Consumer<? super LimitViolation> writeLimitViolation(Network network, TableFormatter formatter) {
+        return violation -> {
+            try {
+                formatter.writeEmptyCell()
+                        .writeEmptyCell()
+                        .writeEmptyCell()
+                        .writeCell(LimitViolation.getCountry(violation, network) != null ? LimitViolation.getCountry(violation, network).name() : "")
```
Comment (mathbagu) [inline security-analysis/src/main/java/com/powsybl/security/Security.java]: LimitViolation.getCountry() cannot return null values: remove this test

File: security-analysis/src/main/java/com/powsybl/security/Security.java [hunk 2]
```diff
@@ -351,16 +575,186 @@ private static LimitViolationKey toKey(LimitViolation violation) {
...
+                        .writeCell(Float.isNaN(LimitViolation.getNominalVoltage(violation, network)) ? "" : Integer.toString((int) LimitViolation.getNominalVoltage(violation, network)))
```
Comment (mathbagu) [inline security-analysis/src/main/java/com/powsybl/security/Security.java]: LimitViolation.getNominalVoltage cannot be NaN

File: security-analysis/src/main/java/com/powsybl/security/Security.java [hunk 3]
```diff
@@ -351,16 +575,186 @@ private static LimitViolationKey toKey(LimitViolation violation) {
...
+                        .writeCell(violation.getSubjectId())
+                        .writeCell(StringUtils.isNotBlank(LimitViolation.getVoltageLevelName(violation, network)) ? LimitViolation.getVoltageLevelName(violation, network) : "")
```
Comment (mathbagu) [inline security-analysis/src/main/java/com/powsybl/security/Security.java]: Use Objects.toString(xxxx, "") instead

File: security-analysis/src/main/java/com/powsybl/security/Security.java [hunk 4]
```diff
@@ -351,16 +575,186 @@ private static LimitViolationKey toKey(LimitViolation violation) {
...
+                        .filter(violation -> preContingencyViolations.isEmpty() || !preContingencyViolations.contains(toKey(violation)))
+                        .collect(Collectors.toList());
+
+                if (!filteredLimitViolations2.isEmpty() || !postContingencyResult.getLimitViolationsResult().isComputationOk()) {
+                    formatter.writeCell(postContingencyResult.getContingency().getId())
+                            .writeCell(postContingencyResult.getLimitViolationsResult().isComputationOk() ? "converge" : "diverge")
+                            .writeEmptyCell()
+                            .writeEmptyCell()
+                            .writeEmptyCell()
+                            .writeCell("(" + filteredLimitViolations2.size() + ")")
```
Comment (mathbagu) [inline security-analysis/src/main/java/com/powsybl/security/Security.java]: Seems strange to have parenthesis here

File: security-analysis/src/main/java/com/powsybl/security/Security.java [hunk 5]
```diff
@@ -351,16 +575,186 @@ private static LimitViolationKey toKey(LimitViolation violation) {
...
+
+            try (TableFormatter formatter = formatterFactory.create(writer,
+                "Post-contingency limit violations",
+                formatterConfig,
+                new Column(CONTINGENCY),
+                new Column(STATUS),
+                new Column(ACTION),
+                new Column(COUNTRY),
+                new Column(BASE_VOLTAGE)
+                    .setHorizontalAlignment(HorizontalAlignment.RIGHT),
```
Comment (mathbagu) [inline security-analysis/src/main/java/com/powsybl/security/Security.java]: country and base voltage columns are related to the voltage level. Move these columns after the "END" column

File: security-analysis/src/main/java/com/powsybl/security/Security.java [hunk 6]
```diff
@@ -301,6 +509,18 @@ private static float getViolationValue(LimitViolation violation) {
...
+        numberFormat.setMinimumFractionDigits(minimumFractionDigits);
+        numberFormat.setMaximumFractionDigits(maximumFractionDigits);
+        numberFormat.setGroupingUsed(false);
+        return numberFormat;
+    }
+
+    /**
+     * @deprecated Use getFormatter(Locale, int, int) instead.
+     */
+    @Deprecated
```
Comment (mathbagu) [inline security-analysis/src/main/java/com/powsybl/security/Security.java]: Remove this method: it's private

File: security-analysis/src/main/java/com/powsybl/security/Security.java [hunk 7]
```diff
@@ -219,14 +332,105 @@ public static String printLimitsViolations(List<LimitViolation> violations, Limi
...
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    private static Consumer<? super LimitViolation> writeLinePreContingencyViolations(Network network, TableFormatter formatter) {
+        return violation -> {
+            try {
+                formatter.writeEmptyCell()
+                        .writeCell(LimitViolation.getCountry(violation, network) != null ? LimitViolation.getCountry(violation, network).name() : "")
```
Comment (mathbagu) [inline security-analysis/src/main/java/com/powsybl/security/Security.java]: The return value of LimitViolation.getCountry(violation, network) cannot be null

File: security-analysis/src/main/java/com/powsybl/security/Security.java [hunk 8]
```diff
@@ -219,14 +332,105 @@ public static String printLimitsViolations(List<LimitViolation> violations, Limi
...
+                        .writeCell(Float.isNaN(LimitViolation.getNominalVoltage(violation, network)) ? "" : Integer.toString((int) LimitViolation.getNominalVoltage(violation, network)))
```
Comment (mathbagu) [inline security-analysis/src/main/java/com/powsybl/security/Security.java]: The return value of LimitViolation.getNominalVoltage(violation, network) cannot be NaN

File: security-analysis/src/main/java/com/powsybl/security/Security.java [hunk 9]
```diff
@@ -219,14 +332,105 @@ public static String printLimitsViolations(List<LimitViolation> violations, Limi
...
+                        .writeCell(violation.getSubjectId())
+                        .writeCell(StringUtils.isNotBlank(LimitViolation.getVoltageLevelName(violation, network)) ? LimitViolation.getVoltageLevelName(violation, network) : "")
```
Comment (mathbagu) [inline security-analysis/src/main/java/com/powsybl/security/Security.java]: Use Objects.toString(xxx, "") instead

File: security-analysis/src/main/java/com/powsybl/security/Security.java [hunk 10]
```diff
@@ -219,14 +332,105 @@ public static String printLimitsViolations(List<LimitViolation> violations, Limi
...
+                ? limitViolationFilter.apply(result.getPreContingencyResult().getLimitViolations(), network)
+                : result.getPreContingencyResult().getLimitViolations();
+
+        try (TableFormatter formatter = formatterFactory.create(writer,
+                "Pre-contingency violations",
+                formatterConfig,
+                new Column(ACTION),
+                new Column(COUNTRY),
+                new Column(BASE_VOLTAGE)
+                    .setHorizontalAlignment(HorizontalAlignment.RIGHT),
```
Comment (mathbagu) [inline security-analysis/src/main/java/com/powsybl/security/Security.java]: The country and base_voltage columns are related to the voltage level: move the columns after the END column

File: security-analysis/src/main/java/com/powsybl/security/Security.java [hunk 11]
```diff
@@ -219,14 +332,105 @@ public static String printLimitsViolations(List<LimitViolation> violations, Limi
     }
 
     private static float getAbsValueLimit(LimitViolation violation) {
-        return Float.valueOf(Float.toString(violation.getLimit()) + (violation.getLimitReduction() != 1f ? " * " + violation.getLimitReduction() : ""));
+        return Math.abs(violation.getValue() - violation.getLimit() * violation.getLimitReduction());
```
Comment (mathbagu) [inline security-analysis/src/main/java/com/powsybl/security/Security.java]: It's strange to use the limitReduction here... The header of the column is "abs(value-limit)" and this value could be misunderstood

File: security-analysis/src/main/java/com/powsybl/security/Security.java [hunk 12]
```diff
@@ -174,21 +284,20 @@ public static String printLimitsViolations(List<LimitViolation> violations, Limi
...
                 "",
                 formatterConfig,
-                new Column("Country"),
-                new Column("Base voltage"),
-                new Column("Equipment (" + filteredViolations.size() + ")"),
-                new Column("Violation type"),
-                new Column("Violation name"),
-                new Column("Value")
+                new Column(COUNTRY),
+                new Column(BASE_VOLTAGE),
```
Comment (mathbagu) [inline security-analysis/src/main/java/com/powsybl/security/Security.java]: The country and the base voltage columns are related to the voltage level. Why this table don't show the voltage level. Move these columns after the voltage level name

File: security-analysis/src/main/java/com/powsybl/security/Security.java [hunk 13]
```diff
@@ -148,17 +176,99 @@ public static String printLimitsViolations(Network network) {
...
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+        return writer.toString().trim();
+    }
+
+    private static Consumer<? super LimitViolation> writeLineLimitsViolations(Network network, TableFormatter formatter) {
+        return violation -> {
+            try {
+                formatter.writeCell(LimitViolation.getCountry(violation, network) != null ? LimitViolation.getCountry(violation, network).name() : "")
```
Comment (mathbagu) [inline security-analysis/src/main/java/com/powsybl/security/Security.java]: The return of LimitViolation.getCountry(violation, network) cannot be null

File: security-analysis/src/main/java/com/powsybl/security/Security.java [hunk 14]
```diff
@@ -148,17 +176,99 @@ public static String printLimitsViolations(Network network) {
...
+                         .writeCell(Float.isNaN(LimitViolation.getNominalVoltage(violation, network)) ? "" : Integer.toString((int) LimitViolation.getNominalVoltage(violation, network)))
```
Comment (mathbagu) [inline security-analysis/src/main/java/com/powsybl/security/Security.java]: The return of LimitViolation.getNominalVoltage(violation, network) cannot be NaN

File: security-analysis/src/main/java/com/powsybl/security/Security.java [hunk 15]
```diff
@@ -148,17 +176,99 @@ public static String printLimitsViolations(Network network) {
...
+                         .writeCell(violation.getSubjectId())
+                         .writeCell(StringUtils.isNotBlank(LimitViolation.getVoltageLevelName(violation, network)) ? LimitViolation.getVoltageLevelName(violation, network) : "")
```
Comment (mathbagu) [inline security-analysis/src/main/java/com/powsybl/security/Security.java]: Use Objects.toString(xxx, "") instead

File: security-analysis/src/main/java/com/powsybl/security/Security.java [hunk 16]
```diff
@@ -148,17 +176,99 @@ public static String printLimitsViolations(Network network) {
...
+
+        NumberFormat numberFormat = getFormatter(formatterConfig.getLocale(), 4, 4);
+        NumberFormat percentageFormat = getFormatter(formatterConfig.getLocale(), 2, 2);
+
+        try (TableFormatter formatter = formatterFactory.create(writer,
+                "",
+                formatterConfig,
+                new Column(COUNTRY),
+                new Column(BASE_VOLTAGE)
+                    .setHorizontalAlignment(HorizontalAlignment.RIGHT),
```
Comment (mathbagu) [inline security-analysis/src/main/java/com/powsybl/security/Security.java]: The country and the base voltage are related to the voltage level: move these columns after the END column

File: security-analysis/src/main/java/com/powsybl/security/Security.java [hunk 17]
```diff
@@ -219,14 +332,105 @@ public static String printLimitsViolations(List<LimitViolation> violations, Limi
...
+                                                     LimitViolationFilter limitViolationFilter) {
+        printPreContingencyViolations(result, network, writer, formatterFactory, TableFormatterConfig.load(), limitViolationFilter);
+    }
+
+    public static void printPreContingencyViolations(SecurityAnalysisResult result, Network network, Writer writer, TableFormatterFactory formatterFactory,
+                                                     TableFormatterConfig formatterConfig, LimitViolationFilter limitViolationFilter) {
+        Objects.requireNonNull(result);
+        Objects.requireNonNull(writer);
+        Objects.requireNonNull(formatterFactory);
+        Objects.requireNonNull(formatterConfig);
```
Comment (mathbagu) [inline security-analysis/src/main/java/com/powsybl/security/Security.java]: Network should not be null

File: security-analysis/src/main/java/com/powsybl/security/LimitViolation.java [hunk 18]
```diff
@@ -148,12 +150,38 @@ static Country getCountry(LimitViolation limitViolation, Network network) {
...
             country = bus.getVoltageLevel().getSubstation().getCountry();
         } else {
-            throw new AssertionError("Unexpected identifiable type: " + identifiable.getClass());
+            throw new AssertionError(UNEXPECTED_IDENTIFIABLE_TYPE + identifiable.getClass());
         }
 
         return country;
     }
 
+    static String getVoltageLevelName(LimitViolation limitViolation, Network network) {
```
[the remark this answers is not in the document]
Comment (calinbanc) [reply security-analysis/src/main/java/com/powsybl/security/LimitViolation.java]: done

File: security-analysis/src/main/java/com/powsybl/security/LimitViolation.java [hunk 19]
```diff
@@ -52,7 +54,7 @@ public LimitViolation(String subjectId, LimitViolationType limitType, float limi
     }
 
     /**
-     * @deprecated use LimitViolation(String, LimitViolationType, String, float, float, float) instead.
+     * @deprecated use LimitViolation(String, Lirearguard action mitViolationType, String, float, float, float) instead.
```
[the remark this answers is not in the document]
Comment (calinbanc) [reply security-analysis/src/main/java/com/powsybl/security/LimitViolation.java]: done

File: security-analysis/src/main/java/com/powsybl/security/Security.java [hunk 20]
```diff
@@ -24,13 +24,29 @@
...
 
     private static final String PERMANENT_LIMIT_NAME = "Permanent limit";
 
+    private static final String CONTINGENCY = "Contingency";
+    private static final String STATUS = "Status";
+    private static final String ACTION = "Action";
+    private static final String COUNTRY = "Country";
+    private static final String BASE_VOLTAGE = "Base voltage";
+    private static final String EQUIPMENT = "Equipment";
+    private static final String END = "End";
```
[the remark this answers is not in the document]
Comment (calinbanc) [reply security-analysis/src/main/java/com/powsybl/security/Security.java]: to discuss with Pauline

File: security-analysis/src/main/java/com/powsybl/security/Security.java [hunk 21]
```diff
@@ -39,15 +55,27 @@
     private Security() {
     }
 
+    /**
+     * @deprecated this method could be safely removed, is no more used.
+     */
+    @Deprecated
```
[the remark this answers is not in the document]
Comment (calinbanc) [reply security-analysis/src/main/java/com/powsybl/security/Security.java]: done

File: security-analysis/src/main/java/com/powsybl/security/Security.java [hunk 22]
```diff
@@ -39,15 +55,27 @@
...
     private static Country getCountry(Branch branch, Branch.Side side) {
         return branch.getTerminal(side).getVoltageLevel().getSubstation().getCountry();
     }
 
+    /**
+     * @deprecated this method could be safely removed, is no more used.
+     */
+    @Deprecated
```
[the remark this answers is not in the document]
Comment (calinbanc) [reply security-analysis/src/main/java/com/powsybl/security/Security.java]: done

File: security-analysis/src/main/java/com/powsybl/security/Security.java [hunk 23]
```diff
@@ -39,15 +55,27 @@
...
     private static float getNominalVoltage(Branch branch) {
         return Math.max(branch.getTerminal1().getVoltageLevel().getNominalV(),
                         branch.getTerminal2().getVoltageLevel().getNominalV());
     }
 
+    /**
+     * @deprecated this method could be safely removed, is no more used.
+     */
+    @Deprecated
```
[the remark this answers is not in the document]
Comment (calinbanc) [reply security-analysis/src/main/java/com/powsybl/security/Security.java]: done

File: security-analysis/src/main/java/com/powsybl/security/Security.java [hunk 24]
```diff
@@ -148,17 +176,99 @@ public static String printLimitsViolations(Network network) {
...
     }
 
+    public static String printLimitsViolations(List<LimitViolation> violations, Network network, LimitViolationFilter filter) {
+        return printLimitsViolations(violations, network, filter, TableFormatterConfig.load());
+    }
+
+    public static String printLimitsViolations(List<LimitViolation> violations, Network network, LimitViolationFilter filter, TableFormatterConfig formatterConfig) {
+        Objects.requireNonNull(violations);
+        Objects.requireNonNull(filter);
+        Objects.requireNonNull(formatterConfig);
```
[the remark this answers is not in the document]
Comment (calinbanc) [reply security-analysis/src/main/java/com/powsybl/security/Security.java]: done

File: security-analysis/src/test/java/com/powsybl/security/SecurityTest.java [hunk 25]
```diff
@@ -38,119 +38,120 @@
...
     public void setUp() {
-        NumberFormat numberFormat = NumberFormat.getInstance(Locale.US);
-        numberFormat.setMaximumFractionDigits(4);
-        numberFormat.setMinimumFractionDigits(4);
-        numberFormat.setGroupingUsed(false);
-
         formatterConfig = new TableFormatterConfig(Locale.US, ',', "inv", true, true);
         // create pre-contingency results, just one violation on line1
-        line1Violation = new LimitViolation("line1", LimitViolationType.CURRENT, 1000f, "20'", 1100);
+        line1Violation = new LimitViolation("NHV1_NHV2_1", LimitViolationType.CURRENT, "Permanent limit", 1000f, 20, 1100, Branch.Side.ONE);
```
[the remark this answers is not in the document]
Comment (calinbanc) [reply security-analysis/src/test/java/com/powsybl/security/SecurityTest.java]: done

File: security-analysis/src/test/java/com/powsybl/security/SecurityTest.java [hunk 26]
```diff
@@ -38,119 +38,120 @@
...
+            .beginTemporaryLimit()
+                .setName("20'")
+                .setAcceptableDuration(20 * 60)
+                .setValue(1200)
+            .endTemporaryLimit()
+            .add();
+        network.getLine("NHV1_NHV2_2").newCurrentLimits2().setPermanentLimit(500f).add();
+
+        return network;
+    }
```
[the remark this answers is not in the document]
Comment (calinbanc) [reply security-analysis/src/test/java/com/powsybl/security/SecurityTest.java]: Use directly the method TestingNetworkFactory.createFromEurostag. Done.

File: security-analysis/src/test/java/com/powsybl/security/SecurityTest.java [hunk 27]
```diff
@@ -38,119 +38,120 @@
...
         PostContingencyResult postContingencyResult = new PostContingencyResult(contingency1, true, Arrays.asList(line1Violation, line2Violation), Collections.singletonList("action2"));
         result = new SecurityAnalysisResult(preContingencyResult, Collections.singletonList(postContingencyResult));
+
+        network = createNetwork();
     }
 
     @Test
     public void printPreContingencyViolations() throws Exception {
         StringWriter writer = new StringWriter();
         //System.out.println(formatterConfig);
```
[the remark this answers is not in the document]
Comment (calinbanc) [reply security-analysis/src/test/java/com/powsybl/security/SecurityTest.java]: done

File: security-analysis/src/main/java/com/powsybl/security/Security.java [hunk 28]
```diff
@@ -405,6 +799,10 @@ public static void printPostContingencyViolations(SecurityAnalysisResult result,
         }
     }
 
+    /**
+     * @deprecated Use writePostContingencyResult(Network, LimitViolationFilter, Set<LimitViolationKey>, TableFormatter) instead.
+     */
+    @Deprecated
```
[the remark this answers is not in the document]
Comment (calinbanc) [reply security-analysis/src/main/java/com/powsybl/security/Security.java]: I can't remove  the method. It is still use by a @Deprecated public method  

/**
     * @deprecated Use printPostContingencyViolations(Network, SecurityAnalysisResult, Writer, TableFormatterFactory, TableFormatterConfig, LimitViolationFilter, boolean) instead.
     */
    @Deprecated
    public static void printPostContingencyViolations(SecurityAnalysisResult result, Writer writer, TableFormatterFactory formatterFactory,
                                                      TableFormatterConfig formatterConfig, LimitViolationFilter limitViolationFilter, boolean filterPreContingencyViolations)

. I'll remove it after I'll clean up the code in ipst.

File: security-analysis/src/main/java/com/powsybl/security/Security.java [hunk 29]
```diff
@@ -351,16 +575,186 @@ private static LimitViolationKey toKey(LimitViolation violation) {
...
+                            .sorted(Comparator.comparing(LimitViolation::getSubjectId))
+                            .forEach(writeLimitViolation(network, formatter));
+                }
+            } catch (IOException e) {
+                throw new UncheckedIOException(e);
+            }
+        };
+    }
+
+    private static Consumer<? super LimitViolation> writeLimitViolation(Network network, TableFormatter formatter) {
```
[the remark this answers is not in the document]
Comment (calinbanc) [reply security-analysis/src/main/java/com/powsybl/security/Security.java]: There are two methods writeLimitViolation. One of them is @Deprecated. I can't adapt the new method to the old method.

File: security-analysis/src/main/java/com/powsybl/security/Security.java (hunk 1 above)
Comment (calinbanc) [reply security-analysis/src/main/java/com/powsybl/security/Security.java]: done

File: security-analysis/src/main/java/com/powsybl/security/Security.java (hunk 2 above)
Comment (calinbanc) [reply security-analysis/src/main/java/com/powsybl/security/Security.java]: I've removed the test Float.isNaN on the LimitViolation.getNominalVoltage. Done.

File: security-analysis/src/main/java/com/powsybl/security/Security.java (hunk 3 above)
Comment (calinbanc) [reply security-analysis/src/main/java/com/powsybl/security/Security.java]: done

File: security-analysis/src/main/java/com/powsybl/security/Security.java (hunk 5 above)
Comment (calinbanc) [reply security-analysis/src/main/java/com/powsybl/security/Security.java]: done

File: security-analysis/src/main/java/com/powsybl/security/Security.java (hunk 6 above)
Comment (calinbanc) [reply security-analysis/src/main/java/com/powsybl/security/Security.java]: I can't. The method is still use by a @Deprecated public method. I'll do it after I'll clean up the code in the ipts.

File: security-analysis/src/main/java/com/powsybl/security/Security.java (hunk 7 above)
Comment (calinbanc) [reply security-analysis/src/main/java/com/powsybl/security/Security.java]: I've removed the test for null. Done.

File: security-analysis/src/main/java/com/powsybl/security/Security.java (hunk 8 above)
Comment (calinbanc) [reply security-analysis/src/main/java/com/powsybl/security/Security.java]: I've removed the test for NaN. Done.

File: security-analysis/src/main/java/com/powsybl/security/Security.java (hunk 9 above)
Comment (calinbanc) [reply security-analysis/src/main/java/com/powsybl/security/Security.java]: done

File: security-analysis/src/main/java/com/powsybl/security/Security.java (hunk 10 above)
Comment (calinbanc) [reply security-analysis/src/main/java/com/powsybl/security/Security.java]: done

File: security-analysis/src/main/java/com/powsybl/security/Security.java (hunk 11 above)
Comment (calinbanc) [reply security-analysis/src/main/java/com/powsybl/security/Security.java]: Done. I've fixed the code.

File: security-analysis/src/main/java/com/powsybl/security/Security.java (hunk 12 above)
Comment (calinbanc) [reply security-analysis/src/main/java/com/powsybl/security/Security.java]: done

File: security-analysis/src/main/java/com/powsybl/security/Security.java (hunk 13 above)
Comment (calinbanc) [reply security-analysis/src/main/java/com/powsybl/security/Security.java]: done

File: security-analysis/src/main/java/com/powsybl/security/Security.java (hunk 14 above)
Comment (calinbanc) [reply security-analysis/src/main/java/com/powsybl/security/Security.java]: Done. I've have removed the test for NaN.

File: security-analysis/src/main/java/com/powsybl/security/Security.java (hunk 15 above)
Comment (calinbanc) [reply security-analysis/src/main/java/com/powsybl/security/Security.java]: done

File: security-analysis/src/main/java/com/powsybl/security/Security.java (hunk 16 above)
Comment (calinbanc) [reply security-analysis/src/main/java/com/powsybl/security/Security.java]: done

File: security-analysis/src/main/java/com/powsybl/security/Security.java (hunk 17 above)
Comment (calinbanc) [reply security-analysis/src/main/java/com/powsybl/security/Security.java]: done

**Pull request merged**